extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -278,19 +278,30 @@ export class AmpA4A extends AMP.BaseElement {
           return Promise.reject(cancellation());
         }
         if (fetchResponse && fetchResponse.arrayBuffer) {
-          return fetchResponse.arrayBuffer().then(
-            bytes => {
-              if (promiseId != this.promiseId_) {
-                return Promise.reject(cancellation());
-              }
-              return this.validateAdResponse_(fetchResponse, bytes)
-                .then(valid => {
+          return fetchResponse.arrayBuffer().then(bytes => {
+            if (promiseId != this.promiseId_) {
+              return Promise.reject(cancellation());
+            }
+            return this.extractCreativeAndSignature(
+                bytes, fetchResponse.headers).then(responseParts => {
                   if (promiseId != this.promiseId_) {
                     return Promise.reject(cancellation());
                   }
-                  return this.maybeRenderAmpAd_(valid, bytes);
+                  return this.validateAdResponse_(
+                      responseParts.creative, responseParts.signature)
+                      .then(valid => {
```

Comment:
I think we can change the API a bit to completely mitigate this. Instead of returning a fulfilled promise (that is, the promise will execute `#then` blocks and not `#catch` blocks) of some boolean value, we can instead either fulfill with the validated creative, or reject with an error. With that, we can rewrite as:

```js
.then(fetchResponse => {
  return fetchResponse.arrayBuffer().then(bytes => {
    return this.extractCreativeAndSignature(bytes, fetchResponse.headers);
  });
}).then(responseParts => {
  return this.validateAdResponse_(responseParts.creative, responseParts.signature);
}).then(creative => {
  // validation succeeded.
  // Importantly, we do not have access to the full response, only the validated creative.
}, error => {
  // validation failed
});